src/test/java/net/minecraftforge/test/TestGuiCapability.java

```diff
@@ -0,0 +1,308 @@
+        {
+            return new TileTestBlock();
+        }
+
+        @Override
+        public boolean onBlockActivated(World worldIn, BlockPos pos, IBlockState state, EntityPlayer playerIn, EnumHand hand,
+                @Nullable ItemStack heldItem, EnumFacing side, float hitX, float hitY, float hitZ)
+        {
+            TileTestBlock tile = (TileTestBlock) worldIn.getTileEntity(pos);
+            IGuiProvider cap = tile.getCapability(CapabilityGuiProvider.GUI_PROVIDER_CAPABILITY, null);
+            playerIn.openGui(cap);
+            return true;
+        }
+    }
+
+    public static class TileTestBlock extends TileEntity
+    {
+
+        TestGuiProvider testGuiProvider;
+
+        public TileTestBlock()
+        {
+            this.testGuiProvider = new TestGuiProvider(this, EnumFacing.UP);
+        }
+
+        @Override
+        public boolean hasCapability(Capability<?> capability, EnumFacing facing)
+        {
+            return capability == CapabilityGuiProvider.GUI_PROVIDER_CAPABILITY || super.hasCapability(capability, facing);
+        }
+
+        @Override public <T> T getCapability(Capability<T> capability, EnumFacing facing)
+        {
+            if(capability == CapabilityGuiProvider.GUI_PROVIDER_CAPABILITY) return (T) testGuiProvider;
+            return super.getCapability(capability, facing);
+        }
+
+        class TestGuiProvider extends GuiProviderTile
+        {
+            public TestGuiProvider(TileEntity tile, EnumFacing side)
+            {
+                super(tile, side);
+            }
+
+            @Override
+            public Object getClientGuiElement(EntityPlayer player, World world, @Nullable TileEntity owner)
+            {
+                return new GuiScreen()
+                {
+                    @Override public void drawScreen(int mouseX, int mouseY, float partialTicks)
+                    {
+                        super.drawScreen(mouseX, mouseY, partialTicks);
+                        drawDefaultBackground();
+                        fontRendererObj.drawString("Gui test", 10, 10, 0xFFFFFF);
+                    }
+                };
+            }
+
+            @Override
+            public Container getServerGuiElement(EntityPlayer player, World world, @Nullable TileEntity owner)
+            {
+                return null;
+            }
+        }
+    }
+
+    private class TestItem extends Item
+    {
+        public TestItem(){
+            setRegistryName("gui", "item");
+            setUnlocalizedName("gui.item");
+            setCreativeTab(CreativeTabs.MISC);
+            setMaxStackSize(1);
+            GameRegistry.register(this);
+        }
+
+        @Override
+        public ActionResult<ItemStack> onItemRightClick(ItemStack itemStackIn, World worldIn, EntityPlayer playerIn, EnumHand hand)
+        {
+            if (itemStackIn.hasCapability(CapabilityGuiProvider.GUI_PROVIDER_CAPABILITY, EnumFacing.UP))
```

Comment:
I still think using `UP` or the hand for the `EnumFacing` is fishy here. Nobody else is going to know how to open your gui. I think `null` should be used, unless you know which side of the block is being clicked.